deepmd/pt/utils/env_mat_stat.py

```diff
@@ -0,0 +1,205 @@
+        zero_mean = torch.zeros(
+            self.descriptor.get_ntypes(),
+            self.descriptor.get_nsel(),
+            4,
+            dtype=env.GLOBAL_PT_FLOAT_PRECISION,
+            device=env.DEVICE,
+        )
+        one_stddev = torch.ones(
+            self.descriptor.get_ntypes(),
+            self.descriptor.get_nsel(),
+            4,
+            dtype=env.GLOBAL_PT_FLOAT_PRECISION,
+            device=env.DEVICE,
+        )
+        for system in data:
+            coord, atype, box, natoms = (
+                system["coord"],
+                system["atype"],
+                system["box"],
+                system["natoms"],
+            )
+            (
+                extended_coord,
+                extended_atype,
+                mapping,
+                nlist,
+            ) = extend_input_and_build_neighbor_list(
+                coord,
+                atype,
+                self.descriptor.get_rcut(),
+                self.descriptor.get_sel(),
+                distinguish_types=self.descriptor.distinguish_types(),
+                box=box,
+            )
+            env_mat, _, _ = prod_env_mat_se_a(
+                extended_coord,
+                nlist,
+                atype,
+                zero_mean,
+                one_stddev,
+                self.descriptor.get_rcut(),
+                # TODO: export rcut_smth from DescriptorBlock
+                self.descriptor.rcut_smth,
+            )
+            env_mat = env_mat.view(
+                coord.shape[0], coord.shape[1], self.descriptor.get_nsel(), 4
+            )
+            env_mats = {}
+
+            if "real_natoms_vec" not in system:
+                end_indexes = torch.cumsum(natoms[0, 2:], 0)
+                start_indexes = torch.cat(
+                    [
+                        torch.zeros(1, dtype=torch.int32, device=env.DEVICE),
+                        end_indexes[:-1],
+                    ]
+                )
+                for type_i in range(self.descriptor.get_ntypes()):
+                    dd = env_mat[
+                        :, start_indexes[type_i] : end_indexes[type_i], :, :
+                    ]  # all descriptors for this element
+                    env_mats[f"r_{type_i}"] = dd[:, :, :, :1]
+                    env_mats[f"a_{type_i}"] = dd[:, :, :, 1:]
+                    yield self.compute_stat(env_mats)
+            else:
+                for frame_item in range(env_mat.shape[0]):
+                    dd_ff = env_mat[frame_item]
+                    atype_frame = atype[frame_item]
+                    for type_i in range(self.descriptor.get_ntypes()):
+                        type_idx = atype_frame == type_i
+                        dd = dd_ff[type_idx]
+                        dd = dd.reshape([-1, 4])  # typen_atoms * nnei, 4
+                        env_mats[f"r_{type_i}"] = dd[:, :1]
+                        env_mats[f"a_{type_i}"] = dd[:, 1:]
+                        yield self.compute_stat(env_mats)
```

Comment:
I am worry about the efficiency of this part of code. 
actually we can use type mask to process across frames in parallel, which is supposed to be more efficient than the for loop. 

The efficiency issue may be left for a future PR.